### PR TITLE
proxy protocol: set downstreamRemoteAddress on StreamInfo

### LINKED ISSUE #14087: Wrong %DOWNSTREAM_REMOTE_ADDRESS% logged when using proxy-protocol with tcp-proxy
When using proxy-protocol as a listener filter, tcp-proxy as the network filter, and access logging, the wrong downstream address is logged. The logged address should be the one from the proxy protocol header on the connection, but the direct peer address is logged instead.

### PR DESCRIPTION
**Pull request by ggreenway**
This fixes a regression which resulted in the downstreamRemoteAddress
on the StreamInfo for a connection not having the address supplied by
the proxy protocol filter, but instead having the address of the
directly connected peer.

This issue does not affect HTTP filters.

Fixes #14087

Signed-off-by: Greg Greenway <ggreenway@apple.com>

Commit Message:
Additional Description:
Risk Level: Low
Testing: New test added to prevent future regressions
Docs Changes: none needed
Release Notes: added
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/server/connection_handler_impl.cc [hunk 1]
```diff
@@ -460,9 +460,10 @@ void ConnectionHandlerImpl::ActiveTcpListener::resumeListening() {
 
 void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
     Network::ConnectionSocketPtr&& socket, std::unique_ptr<StreamInfo::StreamInfo> stream_info) {
-  // Refresh local address in case it was restored by a listener filter like the original_dst
-  // filter.
+  // Refresh addresses in case they are modified by listener filters, such as proxy protocol or
+  // original_dst.
   stream_info->setDownstreamLocalAddress(socket->localAddress());
+  stream_info->setDownstreamRemoteAddress(socket->remoteAddress());
```
Comment (mattklein123) [inline source/server/connection_handler_impl.cc:466]: Can you cut a follow up ticket about auditing this code elsewhere? It think there may be code in the HCM we don't or shouldn't need assuming stream info chaining is working correctly?

**Review by mattklein123 — DISMISSED** [review]
Thanks LGTM

File: source/server/connection_handler_impl.cc (hunk 1 above)
Comment (ggreenway) [reply source/server/connection_handler_impl.cc:466]: #14133

File: test/integration/proxy_proto_integration_test.cc [hunk 2]
```diff
@@ -201,4 +221,59 @@ TEST_P(ProxyProtoIntegrationTest, ClusterProvided) {
...
+  ASSERT_TRUE(fake_upstream_connection->waitForData(5));
+  ASSERT_TRUE(fake_upstream_connection->close());
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
+
+  std::string log_result;
+  // Access logs only get flushed to disk periodically, so poll until the log is non-empty
+  do {
+    log_result = api_->fileSystem().fileReadToEnd(access_log_path);
+  } while (log_result.empty());
```
Comment (mattklein123) [inline test/integration/proxy_proto_integration_test.cc:274]: This should probably have some type of sleep but I guess we have similar issues in other tests where you copied this from? We can fix that some other time.

**Review by mattklein123 — APPROVED** [review]

File: test/integration/proxy_proto_integration_test.cc (hunk 2 above)
Comment (ggreenway) [reply test/integration/proxy_proto_integration_test.cc:274]: Yeah, that part is copy/paste. Although I think I added that originally.

**Comment by hnasr** [comment]
> This fixes a regression which resulted in the downstreamRemoteAddress
> on the StreamInfo for a connection not having the address supplied by
> the proxy protocol filter, but instead having the address of the
> directly connected peer.

Good catch guys,

Was the downstreamRemoteAddress incorrectly reporting Envoy’s source address instead of the downstream client’s?

**Comment by abergmann** [comment]
[CVE-2020-35470](https://nvd.nist.gov/vuln/detail/CVE-2020-35470) was assigned to this request.

**Pull request merged**